api/notification_service.proto

```diff
@@ -0,0 +1,197 @@
+// limitations under the License.
+
+syntax = "proto3";
+
+import "github.com/envoyproxy/protoc-gen-validate/validate/validate.proto";
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "github.com/TheThingsIndustries/protoc-gen-go-json/annotations.proto";
+import "google/api/annotations.proto";
+import "google/protobuf/any.proto";
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+import "lorawan-stack/api/identifiers.proto";
+
+package ttn.lorawan.v3;
+
+option go_package = "go.thethings.network/lorawan-stack/v3/pkg/ttnpb";
+
+// TODO: Migrate away from GoGo Protobuf (https://github.com/TheThingsNetwork/lorawan-stack/issues/2798).
+option (gogoproto.goproto_registration) = true;
+
+message Notification {
+  // The immutable ID of the notification. Generated by the server.
+  string id = 1;
+
+  // The time when the notification was triggered.
+  google.protobuf.Timestamp created_at = 2;
+
+  // The entity this notification is about.
+  EntityIdentifiers entity_ids = 3;
+
+  // The type of this notification.
+  string notification_type = 4;
+
+  // The data related to the notification.
+  google.protobuf.Any data = 5;
+
+  // If the notification was triggered by a user action, this contains the identifiers of the user that triggered the notification.
+  UserIdentifiers sender_ids = 6;
+
+  // If the notification was triggered by an event pipeline, this contains the identifiers of that pipeline.
+  // EventPipelineIdentifiers event_pipeline_ids = 7;
+  reserved "event_pipeline_ids"; reserved 7;
+
+  // Relation of the notification receiver to the entity.
+  repeated NotificationReceiver receivers = 8;
+
+  // Whether an email was sent for the notification.
+  bool email = 9;
+
+  // The status of the notification.
+  NotificationStatus status = 10;
+
+  // The time when the notification status was updated.
+  google.protobuf.Timestamp status_updated_at = 11;
+}
+
+enum NotificationReceiver {
+  option (thethings.json.enum) = { marshal_as_string: true, prefix: "NOTIFICATION_RECEIVER" };
+
+  NOTIFICATION_RECEIVER_UNKNOWN = 0;
+
+  // Notification is received by collaborators of the entity.
+  // If the collaborator is an organization, the notification is received by organization members.
+  NOTIFICATION_RECEIVER_COLLABORATOR = 1;
+
+  // Notification is received by administrative contact of the entity.
+  // If this is an organization, the notification is received by organization members.
+  NOTIFICATION_RECEIVER_ADMINISTRATIVE_CONTACT = 3;
+
+  // Notification is received by technical contact of the entity.
+  // If this is an organization, the notification is received by organization members.
+  NOTIFICATION_RECEIVER_TECHNICAL_CONTACT = 4;
+}
+
+enum NotificationStatus {
+  option (thethings.json.enum) = { marshal_as_string: true, prefix: "NOTIFICATION_STATUS" };
+
+  NOTIFICATION_STATUS_UNSEEN = 0;
+  NOTIFICATION_STATUS_SEEN = 1;
+  NOTIFICATION_STATUS_ARCHIVED = 2;
```

Comment:
So we will only allow to archive and not delete notifications?

Reply:
For the time being, yes.